File: etl/lib/etl/derivations.rb
require "digest"

module Etl
  module Derivations
    module_function

    # Stable, non-reversible per-source-file identifier. Two runs against the
    # same file (same year + filename) always produce the same id, but the
    # id cannot be used to recover the original filename/name.
    def pseudonymous_id(year, filename, salt)
      Digest::SHA256.hexdigest("#{salt}:#{year}:#{filename}")[0, 12]
    end

    # Buckets ages using the same bands already used in the report template's
    # body-fat table (20-39 / 40-59 / 60-79), with catch-alls at the edges.
    def age_range(age)
      return nil if age.nil?

      case age
      when 0..19 then "Under 20"
      when 20..39 then "20-39"
      when 40..59 then "40-59"
      when 60..79 then "60-79"
      else "80+"
      end
    end

    # Blood pressure category, following the Green/Amber/Red thresholds
    # printed in the report ("Cardiovascular Health Measures" section).
    # Uses systolic/diastolic categories.
    def blood_pressure_category(systolic, diastolic)
      return nil if systolic.nil? || diastolic.nil?

      systolic_cat = if systolic >= 140 then :red
                     elsif systolic >= 130 then :amber
                     else :green
                     end
      diastolic_cat = if diastolic >= 90 then :red
                      elsif diastolic >= 85 then :amber
                      else :green
                      end
      worse_category(systolic_cat, diastolic_cat)
    end

    # Non-HDL ("bad") cholesterol category, following the report's table.
    def non_hdl_cholesterol_category(value)
      return nil if value.nil?

      if value > 4.9 then "High"
      elsif value >= 3.37 then "Borderline High"
      else "Optimal"
      end
    end

    # HDL ("good") cholesterol category, following the report's table.
    # Thresholds differ slightly by gender; falls back to male thresholds
    # when gender is unknown.
    def hdl_cholesterol_category(value, gender)
      return nil if value.nil?

      low_threshold = gender.to_s.casecmp("Female").zero? ? 1.29 : 1.03
      if value > 1.55 then "Optimal"
      elsif value >= low_threshold then "Average"
      else "Low"
      end
    end

    # Q4: "good cholesterol, good blood pressure" while otherwise healthy.
    def overall_healthy?(bp_category, non_hdl_category, hdl_category)
      bp_category == "Green" &&
        non_hdl_category != "High" &&
        hdl_category != "Low"
    end

    # Keyword-only detection of nutritional underfuelling language in the
    # Nutrition narrative paragraph (Q4). Best-effort: the source docx has no
    # dedicated structured field for this, only free text.
    UNDERFUELLING_KEYWORDS = [
      "underfuel", "under-fuel", "under fuel",
      "under eating", "undereating", "under-eating",
      "not eating enough", "skipping meals", "skips meals", "skip meals",
      "restrict", "low calorie", "low-calorie", "insufficient intake",
      "not fuelling", "not fueling", "inadequate intake"
    ].freeze

    def nutritional_underfuelling_flag(nutrition_text)
      return false if nutrition_text.nil? || nutrition_text.empty?

      text = nutrition_text.downcase
      UNDERFUELLING_KEYWORDS.any? { |kw| text.include?(kw) }
    end

    def worse_category(a, b)
      order = { green: 0, amber: 1, red: 2 }
      order[a] >= order[b] ? label_for(a) : label_for(b)
    end
    private_class_method :worse_category

    def label_for(sym)
      { green: "Green", amber: "Amber", red: "Red" }[sym]
    end
    private_class_method :label_for
  end
end

File: desktop/etl/derivations.js
// Node.js port of ../../etl/lib/etl/derivations.rb.
"use strict";

const crypto = require("node:crypto");

function pseudonymousId(year, filename, salt) {
  return crypto
    .createHash("sha256")
    .update(`${salt}:${year}:${filename}`)
    .digest("hex")
    .slice(0, 12);
}

function ageRange(age) {
  if (age == null) return null;
  if (age <= 19) return "Under 20";
  if (age <= 39) return "20-39";
  if (age <= 59) return "40-59";
  if (age <= 79) return "60-79";
  return "80+";
}

const CATEGORY_ORDER = { green: 0, amber: 1, red: 2 };
const CATEGORY_LABEL = { green: "Green", amber: "Amber", red: "Red" };

function worseCategory(a, b) {
  return CATEGORY_ORDER[a] >= CATEGORY_ORDER[b] ? CATEGORY_LABEL[a] : CATEGORY_LABEL[b];
}

function bloodPressureCategory(systolic, diastolic) {
  if (systolic == null || diastolic == null) return null;

  const systolicCat = systolic >= 140 ? "red" : systolic >= 130 ? "amber" : "green";
  const diastolicCat = diastolic >= 90 ? "red" : diastolic >= 85 ? "amber" : "green";
  return worseCategory(systolicCat, diastolicCat);
}

function nonHdlCholesterolCategory(value) {
  if (value == null) return null;
  if (value > 4.9) return "High";
  if (value >= 3.37) return "Borderline High";
  return "Optimal";
}

function hdlCholesterolCategory(value, gender) {
  if (value == null) return null;
  const lowThreshold = (gender || "").toLowerCase() === "female" ? 1.29 : 1.03;
  if (value > 1.55) return "Optimal";
  if (value >= lowThreshold) return "Average";
  return "Low";
}

function overallHealthy(bpCategory, nonHdlCategory, hdlCategory) {
  return bpCategory === "Green" && nonHdlCategory !== "High" && hdlCategory !== "Low";
}

const UNDERFUELLING_KEYWORDS = [
  "underfuel", "under-fuel", "under fuel",
  "under eating", "undereating", "under-eating",
  "not eating enough", "skipping meals", "skips meals", "skip meals",
  "restrict", "low calorie", "low-calorie", "insufficient intake",
  "not fuelling", "not fueling", "inadequate intake"
];

function nutritionalUnderfuellingFlag(nutritionText) {
  if (!nutritionText) return false;
  const text = nutritionText.toLowerCase();
  return UNDERFUELLING_KEYWORDS.some((kw) => text.includes(kw));
}

module.exports = {
  pseudonymousId,
  ageRange,
  bloodPressureCategory,
  nonHdlCholesterolCategory,
  hdlCholesterolCategory,
  overallHealthy,
  nutritionalUnderfuellingFlag,
  UNDERFUELLING_KEYWORDS
};
